NaNs in classification heads upon checkout + init

### System Info

software
`torch=2.12.0` `transformers=5.8.1`
hardware 
`NVIDIA-SMI 580.82.07              Driver Version: 580.82.07      CUDA Version: 13.0`  


### Who can help?

call for help
@CyrilVallez - as model loading and init seems wrong
@stevhliu - as this can be better docummented

thanks!

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

Hey, so I am seeing many models from `AutoModelForSequenceClassification` 
initialised with NaNs in their heads, 
so that the training can't continue. 

One workaround is to force `float32` at checkout, which is non-default and not indended.

```python
model = AutoModelForSequenceClassification.from_pretrained(
    'microsoft/deberta-v3-base', 
    num_labels=2,   
    dtype=torch.float32, # FIXME seems to be only working with this
    # _fast_init=True,
)

model.to("cuda:0")

nan_params = [name for name, param in model.named_parameters() if param.isnan().any()]
assert len(nan_params) == 0, f"NaN weights in: {nan_params}"

# AssertionError: NaN weights in: ['pooler.dense.weight', 'classifier.weight']
```

### Expected behavior

initialised with numerical weights - no nans - from checkout, under default settings

--- a/src/transformers/modeling_utils.py
+++ b/src/transformers/modeling_utils.py
@@ -2400,7 +2400,7 @@ class PreTrainedModel(nn.Module, EmbeddingAccessMixin, ModuleUtilsMixin, PushToH
 
         if isinstance(module, (nn.Linear, nn.Conv1d, nn.Conv2d, nn.Conv3d, nn.ConvTranspose1d, nn.ConvTranspose2d)):
             if getattr(module, "weight", None) is not None:
-                init.normal_(module.weight.float(), mean=0.0, std=std)
+                init.normal_(module.weight, mean=0.0, std=std)
             if module.bias is not None:
                 init.zeros_(module.bias)
         elif isinstance(module, nn.Embedding):

